### DISCUSSION TITLE
Project roadmap ideas

### DISCUSSION THREAD

**Discussion by kaveh808** [kaveh808/kons-9#101] [Ideas]
[Moving this from another discussion.]

The goal of the project is to develop a platform -- an IDE for 3D graphics -- on which can be built all of the above and more.

My personal interests and early development tend towards procedural generation, for which I find CL well-suited.

In my mind, the real measure of success of the system will be determined by whether a community flocks to it and develops their own code and tools on what we've built.

The key at this stage is to develop an architecture which encourages such development.

I see Maya, Houdini, and Blender as a source of ideas and inspiration, mixed with a bit of own magic sauce: CL.

**Comment by ghost** [comment]
I think this is a good place to start, with respect to a CG production system--from https://www.blender.org/features:

Blender Features

* Rendering
* Modeling
* Sculpting
* Animation and Rigging
* Story Artist
* VFX
* Simulation
* Pipeline
* Video Editing
* Scripting (Python API)
* Interface (GUI)

Where to start?  Which of these do we think Common Lisp would have some advantage?  CAD/CAM is another application area to expand this list.

**Comment by kaveh808** [comment]
My list from another discussion:

cubic parametric curves -- bezier curve example has been implemented, need other curves
implement poly-mesh class -- the class has been implemented, need topological operations for mesh editing
cubic parametric surfaces -- spline patches
trimmed NURBS -- useful for CAD software
subdivision surfaces -- based off poly-mesh, with edge sharpness weighting
csg -- for CAD
signed distance functions -- implicit rendering of mathematical functions
isosurfaces -- for metaballs style modeling and data visualization
genetic algorithms -- general approach to very the parameters of any system
genetic programming -- use genetic techniques to generate code for a system
3d ui manipulators -- interactive controls for transforming shapes
3d ui picking -- ability to pick shapes, faces, edges, and vertices in the 3D view
gui toolkit -- cross platform UI system built on a graphics engine
gui text engine -- ability to display scaled and styled text
computed-slots -- data flow management between objects
persistent objects -- ability to store and load scene data
rigid body dynamics -- possibly wrapper around existing C package
soft body dynamics -- possibly wrapper around existing C package
fluid simulation -- possibly wrapper around existing C package
behaviors -- shapes with a state machine
protocols -- generic interface mechanism for procedural asset creation
deformers -- bend/twist/taper, FFD lattices, etc
deformable-mesh -- adaptively subdividable mesh for deformation
character animation: skeleton rigging
character animation: muscles system
character animation: skinning
character animation: facial animation
character animation: hair and fur
character animation: clothing
I/o: obj
I/o: usd
dsl for modeling
animation: timeline
animation: playback controls
animation: sequencer
animation: dsl for blocking/choreography
gui design: customizable extensible inspectors (aspects)
gui generates code
particles: build on existing system
l-systems: build on existing system
manager-groups: variants
manager-groups: LOD
manager-groups: blend shapes
scene i/o
visual programming
code visualization
batch engine (no ui)
gpu & simd code
collaboration features
standalone app
layered gestural animation
purely key-driven ui
gui for editing code
emergent structures
adaptive ui
grid view for genetic evolution
generate geometric styles
web features
web graphics output
data visualization
mathematical visualization

**Reply by ghost** [reply]
# Long Term Ideas [WIP]

## Modeling

dsl for modeling
particles: build on existing system
l-systems: build on existing system
manager-groups: variants
manager-groups: LOD
manager-groups: blend shapes
emergent structures

### B-Rep modeling

cubic parametric curves -- bezier curve example has been implemented, need other curves
implement poly-mesh class -- the class has been implemented, need topological operations for mesh editing
cubic parametric surfaces -- spline patches
trimmed NURBS -- useful for CAD software
subdivision surfaces -- based off poly-mesh, with edge sharpness weighting
csg -- for CAD
deformers -- bend/twist/taper, FFD lattices, etc
deformable-mesh -- adaptively subdividable mesh for deformation

### F-rep modeling

csg -- for CAD
signed distance functions -- implicit rendering of mathematical functions
isosurfaces -- for metaballs style modeling and data visualization
deformers -- bend/twist/taper, FFD lattices, etc

## Interop
I/o: obj
I/o: usd

## Animation

character animation: skeleton rigging
character animation: muscles system
character animation: skinning
character animation: facial animation
character animation: hair and fur
character animation: clothing
behaviors -- shapes with a state machine
animation: timeline
animation: playback controls
animation: sequencer
animation: dsl for blocking/choreography
layered gestural animation

## Interface (GUI and otherwise)

3d ui manipulators -- interactive controls for transforming shapes
3d ui picking -- ability to pick shapes, faces, edges, and vertices in the 3D view
gui toolkit -- cross platform UI system built on a graphics engine
gui text engine -- ability to display scaled and styled text
gui design: customizable extensible inspectors (aspects)
gui generates code
visual programming
gui for editing code
batch engine (no ui)
purely key-driven ui
adaptive ui
collaboration features
standalone app


## Simulation

fluid simulation -- possibly wrapper around existing C package
genetic algorithms -- general approach to very the parameters of any system
genetic programming -- use genetic techniques to generate code for a system
grid view for genetic evolution


## Common Lisp features

computed-slots -- data flow management between objects
persistent objects -- ability to store and load scene data
protocols -- generic interface mechanism for procedural asset creation
scene i/o
gpu & simd code


## Web

web features
web graphics output


## Visualization

data visualization
mathematical visualization
code visualization


## Unclassified

Generate geometric styles

**Reply by ghost** [reply]
*Long Term*

For CG production, I would add scripted video editing/post production.

https://en.wikipedia.org/wiki/Edit_decision_list

For generative design I would add ML with SDF:

https://openaccess.thecvf.com/content_CVPR_2019/papers/Park_DeepSDF_Learning_Continuous_Signed_Distance_Functions_for_Shape_Representation_CVPR_2019_paper.pdf

**Comment by kaveh808** [comment]
I see us starting with the following functionality:

- polygonal modeling (mainly procedural)
- animation (mainly procedural)
- rendering facility (as file export or FFI call to existing renderer)
- GUI
- save & load scenes

**Reply by foretspaisibles** [reply]
This looks likes an interesting pick to me. I believe rendering is the more crucial one to make the software a useful tool and also makes it possible to implement some kind of high-level testing.

**Reply by ghost** [reply]
The use of the word "rendering" is ambiguous here.  OpenGL is a rendererer as are Filament and LuxCoreRender.  My interpretation of "rendering facility" is the generation of production images, not the rendering during modeling--although Filament blurs the line a little being real-time or near real-time.

**Reply by ghost** [reply]
> I see us starting with the following functionality:

Looks like a reasonable subset to start with.

**Reply by JMC-design** [reply]
and that's an important distinction to make.  Is it important to have foreign renderers doing realtime display within the app, or just having the ability to send scenes to an external renderer, maybe distributed, to produce a final product?  Especially if foreign renderers can take some standard scene description that we can also have as an export format.
I'm not really sold on filament. It's very mobile and gamey looking. While some nice realtime pbr rendering is still possible under lisp if you look at some of mfiano's game engine demos.

**Reply by ghost** [reply]
In the good old days, Ray Dream Studio had (1) simple Phong shading (IIRC) for doing the real-time geometric modeling, (2) a more realistic shader-based render preview for near real-time, and (3) a final rendering output to a production file.  It was sort of a progressive process of realism.  (I'm old enough to remember wireframe modeling in mainframe CAD systems--you had to use your imagination more.) Meshlab has some similar ideas which would be good to look at--including custom shaders and flexible interactive rendering.  Someone also mentioned ShaderToy.

> I'm not really sold on filament.

I'm sold on making as many people happy as possible, even if only we are happy initially. Having 1 - 3 customizable might help more people adapt to their own style and get the technique that suits their preference.  With the soft requirement of attracting more users and devs, this might be worth some extra effort:

https://en.wikipedia.org/wiki/Fundamental_theorem_of_software_engineering

> The fundamental theorem of software engineering is a term...attributed to David J. Wheeler: "We can solve any problem by introducing an extra level of indirection." The theorem does not describe an actual theorem that can be proven; rather, it is a general principle for managing complexity through abstraction.

**Reply by JMC-design** [reply]
I'm old enough to have used 3DS in DOS!
all are doable under lisp.  After all, the majority of the work is just writing shaders and transferring data.

**Reply by ghost** [reply]
> I'm old enough to have used 3DS in DOS!

We can add "sparse point cloud rendering" in the future, if that is what you are used to. (Seriously, I think Meshlab has this).

> all are doable under lisp

I've been looking at some newer hardware designs for doing F-rep rendering, and found out last week that even chip designers use indirection: register renaming.  (Everybody else here probably already knows what that is.)

**Reply by kaveh808** [reply]
We can think of renderers in three categories:
1) 3D modeler display -- e.g. OpenGL
2) real-time renderer -- e.g. Filament
3) batch renderer -- send off files for rendering (e.g. prman)

1 and 2 would be available as windows in our UI.

I suggested Filament to try out because it is real-time, has major backing, and CL bindings already exist. But I'm open to other candidates.

As far as being old enough to... I go back to writing ray tracers in Fortran on VAX/VMS...

**Comment by JMC-design** [comment]
I may be controversial here, but I think things like parsing/writing external formats should be a separate library.
As well as formulaic construction of geometric objects.  These are things that all projects can benefit from.
Same thing as things like shader libraries, simple gathering of things like https://iquilezles.org/articles/distfunctions2d/

**Reply by kaveh808** [reply]
I am not opposed to the idea, once our design has settled down. If all goes well, the most useful such library may be the GUI one.

**Comment by starkaiser** [comment]
I think we should design kons-9 just how Lisp is designed: a small kernel with basic functionalities that can be use to build everything else. I am interested in NURBS, because right now there is no good free software alternative to Rhinoceros 3D and Grasshopper. Kaveh started this project at the right time:). Just before he announced the project I decided to try and implement NURBS in CL so I could use the features of Lisp for algorithmic design. I started working on the implementation, but it will not be usable anytime soon (I am not a programmer, I program in my free time)

> I see us starting with the following functionality:
> 
> - polygonal modeling (mainly procedural)
> - animation (mainly procedural)
> - rendering facility (as file export or FFI call to existing renderer)
> - GUI
> - save & load scenes

This is a good starting point, I think. If the GUI is made first, people working on different things can use the same visualization routines to help them while developing the software and we can have a common API to display the 3D objects, regardless of their representation (polygon mesh / NURBS)

**Reply by bigos** [reply]
Can we add simple things as well? Blender has a text file viewer and image viewer. Could that be a good starting point towards other plugins?

**Reply by JMC-design** [reply]
Image viewer is pretty easy, basically got that already.
Text display is no small task.  Sure it's dead simple to get some text to display, but typography and making it readable are the sticking points.  I'm hoping when my FONT stuff is more tested I'll point it out to Didier and he'll work his paragraph layout stuff with it.  The problem with his software, like almost all CL software, is that it doesn't deal with screen output! So everything is handed off to some foreign system.
Now if you're just talking about displaying mono text with naive outputting like terminals, sure.

**Reply by kaveh808** [reply]
What do the Nyxt browser guys use for text? Is their stuff all CL?

**Reply by JMC-design** [reply]
Nah, it's mainly built around ffi to things like webkit.

**Reply by bigos** [reply]
I do not know to what extent
https://learnopengl.com/In-Practice/Text-Rendering
reflects your plans.

**Reply by bigos** [reply]
I ended up looking at http://freetype.org/, but I am not any wiser.

**Reply by JMC-design** [reply]
@bigos  I'm not sure who you're asking or what you're getting at, but freetype is only a font rasterizer.  This is separate from display which is talked about in the first link you posted.
All of which is separate from paragraph layout where there are different algorithms to calculate spacing and such, which is also separate from, but part of, typography which is all about proper use of text styles, fonts, and layouts to convey meaning and make things readable.

**Reply by kaveh808** [reply]
@bigos We are already working on a text engine which we hope will be ready soon. The initial use will be for the GUI. Stay tuned. :)

**Reply by bigos** [reply]
Definitely, staying tuned is better than reinventing the wheel. And without my own working prototype, it is difficult to explain where I may have been misunderstood. And with your text engine, I may see the problem from a different angle.
